tests/backends/base/test_creation.py

```diff
@@ -73,3 +75,27 @@ def test_migrate_test_setting_true(self, mocked_migrate, mocked_ensure_connectio
         finally:
             with mock.patch.object(creation, '_destroy_test_db'):
                 creation.destroy_test_db(old_database_name, verbosity=0)
+
+
+class TestDeserializeDbFromString(SimpleTestCase):
+    databases = ['default']
+
+    def test_forward_reference(self):
+        """
+        This has the book before the author it references, to verify deserialize_db_from_string handles this properly
+        (regression test for #26552).
+        """
+
+        serialized = """
+        [
+            {"model": "backends.book", "pk": 1, "fields": {"author": "John"}},
+            {"model": "backends.author", "pk": 1, "fields": {"name": "John"}}
+        ]
+        """
```

Comment:
> Normally these objects will appear in a correct order because serialize_db_to_string() sorts dependencies. 

I believe this sorting only applies to models with natural keys, for regular keys the serialization will assume that things run in a transaction and can be inserted in any order. So even without circular references, this problem could already occur. Of course, circular references will guarantee the problem occurs.

> I've changed this to Object and ObjectReference with circular dependencies to create a real-life test.

I had avoided using `ObjectReference`, since its `ForeignKey` has `db_constraint=False` which I assumed would prevent the problem from triggering. Did you try your new testcase without the fix applied?

Reply:
> I had avoided using ObjectReference, since its ForeignKey has db_constraint=False which I assumed would prevent the problem from triggering. Did you try your new testcase without the fix applied?

Yes, it fails without this patch with an `IntegrityError`.

Reply:
Ah, I just checked as well, which is because you have the `Object` (which references *with* db constraints) before the `ObjectReference`. If I swap them, the `IntegrityError` disappears. So the "circular reference so more realistic" argument disappears a bit (but the test-case as it is now works, so I guess that's fine).

Reply:
When you will create an app with `Object` and `ObjectReference` as defined in `backends.models` then `serialize_db_to_string()` generates `data` with order that causes an `IntegrityError`. That is what I meant by a "real-life" example.